### PR TITLE
Android biometric settings

### PR DESCRIPTION
**Pull request by perunt**
Fixes RNBW-1886

## What changed (plus any additional context for devs)
FIxed unavailable phrase list when a user creates a wallet with PIN and after then enable a biometric
Added text that warned the user this account was created with biometric, and he tried to see prases without enabled biometric.

## Dev checklist for QA: what to test
- wallet creation/importing
- look at the recovery phrase while enabling and disabling biometrics. For wallet created with PIN and with biometric

## Final checklist

- [ ] Assigned individual reviewers?
- [ ] Added labels?

### CODE REVIEW

File: src/handlers/authentication.ts [hunk 1]
```diff
@@ -36,7 +36,7 @@ export async function savePIN(pin: any) {
   }
 }
 
-export async function authenticateWithPIN() {
+export async function authenticateWithPIN(safeIfNeed: boolean = false) {
```
Comment (osdnk) [inline src/handlers/authentication.ts:39]: `safeIfNeed`? what do you mean?

File: src/model/wallet.ts [hunk 2]
```diff
@@ -174,6 +174,9 @@ export enum WalletLibraryType {
   ethers = 'ethers',
   bip39 = 'bip39',
 }
+export enum LoadSeedType {
+  CreatedWithBiometricError = 'createdWithBiometricError',
+}
```
Comment (osdnk) [inline src/model/wallet.ts:179]:
```suggestion
const LoadSeedType {
  CreatedWithBiometricError = 'createdWithBiometricError',
} as const
export LoadSeedType
```

We don't like enums

File: src/components/secret-display/SecretDisplaySection.js [hunk 3]
```diff
@@ -128,6 +135,34 @@ export default function SecretDisplaySection({
   const typeLabel = type === WalletTypes.privateKey ? 'key' : 'phrase';
 
   const { colors } = useTheme();
+
+  const renderStep = useCallback(() => {
+    if (isHideRecoveryPhrase) {
+      return (
+        <ColumnWithMargins align="center" justify="center">
+          <AuthenticationText>
+            Your account has been created with biometric data. To see the
```
Comment (osdnk) [inline src/components/secret-display/SecretDisplaySection.js:144]: cc @jeremygoldbrg @christianbaroni @mikearndt @g-a-v-i-n for copy

File: src/components/secret-display/SecretDisplaySection.js [hunk 4]
```diff
@@ -92,12 +93,15 @@ export default function SecretDisplaySection({
...
   const currentWallet = wallets[walletId];
   const [visible, setVisible] = useState(true);
+  const [isHideRecoveryPhrase, setHideRecoveryPhrase] = useState(false);
   const [seed, setSeed] = useState(null);
   const [type, setType] = useState(currentWallet?.type);
 
   const loadSeed = useCallback(async () => {
     try {
-      const s = await loadSeedPhraseAndMigrateIfNeeded(walletId);
+      const { seed: s, actionStatus } = await loadSeedPhraseAndMigrateIfNeeded(
```
Comment (osdnk) [inline src/components/secret-display/SecretDisplaySection.js:102]: this flow is a bit funky. I see what you're doing, but it's hard to follow.
Maybe this is a good moment to just throw an error inside `src/handlers/authentication.ts` and handle it here? Cause this is actually error what happened 

Or maybe something else, just I have a preference for not "replicating" the structure of add data struct just for reducing type-handling

**Review by osdnk — CHANGES_REQUESTED** [review]
see comments, good overall. I just block this until obtaining copy approval. Please, include screenshots / recording.

File: src/handlers/authentication.ts (hunk 1 above)
Comment (perunt) [reply src/handlers/authentication.ts:39]: if I pass this prop, the pin will be saved if we haven't the previous one
I will replace it with `shouldSavePin`

File: src/model/wallet.ts (hunk 2 above)
Comment (perunt) [reply src/model/wallet.ts:179]: Just use strings?

File: src/handlers/authentication.ts [hunk 5]
```diff
@@ -36,7 +36,7 @@ export async function savePIN(pin: any) {
   }
 }
 
-export async function authenticateWithPIN() {
+export async function authenticateWithPIN(shouldSavePin: boolean = false) {
```
Comment (terrysahaidak) [inline src/handlers/authentication.ts:39]: I would use `true` here as default so it won't break it in the other places

File: src/model/wallet.ts [hunk 6]
```diff
@@ -630,7 +630,7 @@ export const createWallet = async (
           if (!userPIN) {
             // We gotta dismiss the modal before showing the PIN screen
             dispatch(setIsWalletLoading(null));
-            userPIN = await authenticateWithPIN();
+            userPIN = await authenticateWithPIN(true);
```
Comment (terrysahaidak) [inline src/model/wallet.ts:633]: So we can remove this change

File: src/model/wallet.ts [hunk 7]
```diff
@@ -1041,7 +1041,7 @@ export const generateAccount = async (
           const { dispatch } = store;
           // Hide the loading overlay while showing the pin auth screen
           dispatch(setIsWalletLoading(null));
-          userPIN = await authenticateWithPIN();
+          userPIN = await authenticateWithPIN(true);
```
Comment (terrysahaidak) [inline src/model/wallet.ts:1044]: And this one too

**Review by terrysahaidak — CHANGES_REQUESTED** [review]

File: src/components/secret-display/SecretDisplaySection.js (hunk 3 above)
Comment (terrysahaidak) [reply src/components/secret-display/SecretDisplaySection.js]: also, I'm pretty sure there is PR for i18n touching this

**Comment by tchayen** [comment]
This touches very tricky parts regarding security (obviously) and UX (we want to avoid situations when we ask user for any credentials before it's necessary) so (probably this is obvious but stating just in case) please don't merge without review from Bruno :pray:.

File: src/components/secret-display/SecretDisplaySection.js [hunk 8]
```diff
@@ -92,6 +92,7 @@ export default function SecretDisplaySection({
   const walletId = params?.walletId || selectedWallet.id;
   const currentWallet = wallets[walletId];
   const [visible, setVisible] = useState(true);
+  const [isRecoveryPhraseVisible, setRecoveryPhraseVisible] = useState(false);
```
Comment (osdnk) [inline src/components/secret-display/SecretDisplaySection.js:95]:
```suggestion
  const [isRecoveryPhraseVisible, setIsRecoveryPhraseVisible] = useState(false);
```

File: src/components/secret-display/SecretDisplaySection.js [hunk 9]
```diff
@@ -106,8 +107,12 @@ export default function SecretDisplaySection({
       }
       setVisible(!!s);
       onSecretLoaded?.(!!s);
+      setRecoveryPhraseVisible(!!s);
     } catch (e) {
       logger.sentry('Error while trying to reveal secret', e);
+      if (e?.message === 'createdWithBiometricError') {
+        setRecoveryPhraseVisible(false);
```
Comment (osdnk) [inline src/components/secret-display/SecretDisplaySection.js:114]:
```suggestion
        setIsRecoveryPhraseVisible(false);
```

File: src/components/secret-display/SecretDisplaySection.js [hunk 10]
```diff
@@ -162,18 +195,7 @@ export default function SecretDisplaySection({
...
-          </AuthenticationText>
-          <ToggleSecretButton onPress={loadSeed}>
-            <BiometricButtonContent
-              color={colors.white}
-              label={`Show Recovery ${upperFirst(typeLabel)}`}
-              showIcon={!seed}
-            />
-          </ToggleSecretButton>
-        </ColumnWithMargins>
+        renderStep()
```
Comment (osdnk) [inline src/components/secret-display/SecretDisplaySection.js:198]:
```suggestion
        renderStepNoSeeds()
```

File: src/components/secret-display/SecretDisplaySection.js [hunk 11]
```diff
@@ -128,6 +133,34 @@ export default function SecretDisplaySection({
   const typeLabel = type === WalletTypes.privateKey ? 'key' : 'phrase';
 
   const { colors } = useTheme();
+
+  const renderStep = useCallback(() => {
```
Comment (osdnk) [inline src/components/secret-display/SecretDisplaySection.js:137]:
```suggestion
  const renderStepNoSeeds = useCallback(() => {
```

File: src/model/wallet.ts [hunk 12]
```diff
@@ -1294,6 +1305,6 @@ export const loadSeedPhraseAndMigrateIfNeeded = async (
   } catch (error) {
     logger.sentry('Error in loadSeedPhraseAndMigrateIfNeeded');
     captureException(error);
-    return null;
+    throw error;
```
Comment (osdnk) [inline src/model/wallet.ts:1308]:
```suggestion
    throw null;
```

**Comment by perunt** [comment]
two types of behavior are covered here
1. the wallet was created with biometrics and biometrics were turned off after then
As the data in the keychain are not accessible unless the biometric is not turned on again we haven't `seedData` and `seedPhrase`. So in that case I just show a warning message.

2. the wallet was created with a PIN and biometrics were turned on after then
We need just to check if `seedPhrase` includes PIN info, get PIN from a user, compare it with the existing PIN and decrypt it

**Comment by brunobar79** [comment]
I'll review before monday

File: src/model/wallet.ts (hunk 12 above)
Comment (perunt) [reply src/model/wallet.ts:1307]: we need to distinguish what error has occurred and throw it up

Comment (osdnk) [reply src/model/wallet.ts:1307]: why this was not needed before?

File: src/handlers/authentication.ts [hunk 13]
```diff
@@ -36,7 +36,7 @@ export async function savePIN(pin: any) {
   }
 }
 
-export async function authenticateWithPIN() {
+export async function authenticateWithPINAndCreateIfNeeded() {
```
Comment (osdnk) [inline src/handlers/authentication.ts:39]: for context: this was renamed, because now it better reflects what is the purpose of the function

**Review by brunobar79 — APPROVED** [review]
@perunt  This looks good to me. Good stuff! (PR description / context / videos 10/10)

Code is approved but please before merging just get some feedback about this screen on #design  so we can make it look a bit better

Screen Shot 2022-04-04 at 5 54 46 PM

File: src/model/wallet.ts (hunk 12 above)
Comment (perunt) [reply src/model/wallet.ts:1307]: We didn't need to determine what type of error occurred, we didn't show a recovery phrase for any of the errors. Now in case of an access error, we need not only to hide the phrase but also to show notifications with instructions, and that throw help to understand when we should show this instruction

**Review by osdnk — APPROVED** [review]
ok, i think I'm good with those changes if @brunobar79 says it's fine

**Review by terrysahaidak — APPROVED** [review]
Looks good enough for me

**Comment by christianbaroni** [comment]
looks like the text in the screenshot is using the wrong font

**Comment by perunt** [comment]
@christianbaroni I used the same style as for this warning, should they both be changed?
photo_2022-04-11 21 21 35
?

**Review by ibrahimtaveras00 — APPROVED** [review]
All looks good in regards to backup/viewing seed phrase or private key

There is an issue that I noticed dealing with confirming txn's that's out of the scope of this PR and already happens on playstore App seen here = RNBW-3259

**Comment by tchayen** [comment]
@perunt @ibrahimtaveras00 
Pinging as this PR is waiting here for a while. I think @perunt can merge it, right?

**Comment by ibrahimtaveras00** [comment]
This passed QA on my end, if there are any changes that needs my eyes again, just add the needs QA label and I'll hop back on it, if not then we should be good whenever ready 👍🏽
